### PR TITLE
rustc_parse: Remove optimization for 0-length streams in `collect_tokens`

### PR DESCRIPTION
**Pull request by petrochenkov**
The optimization conflates empty token streams with unknown token stream, which is at least suspicious, and doesn't affect performance because 0-length token streams are very rare.

r? @Aaron1011

### CODE REVIEW

**Comment by petrochenkov** [comment]
Some more statistics - number of calls to `collect_tokens` and to `create_token_stream` for different values of `num_calls`.
Looks like lazy tokens streams are pretty rarely converted to real token streams.
Figure_1

**Comment by petrochenkov** [comment]
This means that
- we perhaps need to find a way to collect tokens in less situations
- make collection as cheap as possible in the meantime

Right now the construction of `LazyTokenStreamImpl` is free, and this PR introduces 2 more allocations and deallocations for all 1-token streams, which are the most common ones.

**Comment by petrochenkov** [comment]
Marking as blocked on https://github.com/rust-lang/rust/pull/78782, I'll re-collect the statistics after that PR lands.

**Comment by petrochenkov** [comment]
New statistics for collected tokens (0-length token streams are entirely insignificant):
```
num_calls      count
        1     260184
        4      45825
        7      27256
       13      15371
       10      13940
       17      11646
        5      11438
        3      10148
        9       8058
        2       6038
        6       5300
       16       2674
        8       2089
       18       1700
       12       1699
       11       1363
       14       1251
       20       1020
       21        989
       19        857
```

New statistics for token streams created from collected tokens (creation is still much rarer than collection):
```
num_calls      count
        3        312
        4        210
        1        185
       15        108
        7         97
        9         91
       16         82
       17         77
       21         66
       11         62
       14         54
       10         49
       23         45
       26         45
       29         42
       18         41
       25         41
       24         38
       27         38
       28         37
```

Staticstics for `collect_tokens` call sources:
```
source                     count
parse_nonterminal(expr)    328921
parse_attribute            62941
parse_item_common          16896
parse_nonterminal(meta)    15855
parse_nonterminal(ty)      15274
parse_nonterminal(literal) 2633
parse_attr_item            1870
parse_nonterminal(vis)     1032
parse_nonterminal(pat)     649
parse_prefix_expr          512
parse_nonterminal(block)   117
parse_nonterminal(path)    63
```

**Comment by petrochenkov** [comment]
Updated.
Only the 0-length token stream change is left.

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -1247,8 +1243,8 @@ impl<'a> Parser<'a> {
 
         let lazy_impl = LazyTokenStreamImpl {
             start_token,
+            num_calls: self.token_cursor.num_next_calls - cursor_snapshot.num_next_calls,
             cursor_snapshot,
-            num_calls,
             desugar_doc_comments: self.desugar_doc_comments,
         };
         Ok((ret, Some(LazyTokenStream::new(lazy_impl))))
```
Comment (Aaron1011) [inline compiler/rustc_parse/src/parser/mod.rs:1250]: Do you still want `collect_tokens` to return an `Option<LazyTokenStream>`?

Comment (petrochenkov) [reply compiler/rustc_parse/src/parser/mod.rs:1250]: I want to include the
```
let (thing, tokens) = self.collect_tokens(...);
if thing.tokens.is_none() {
    thing.tokens = tokens;
}
```
logic into `collect_tokens` itself instead, but that's probably a task for another PR.

**Comment by Aaron1011** [comment]
spurious zlib inflation error

@bors retry

**Pull request merged**